### PR TITLE
Store and retrieve "Implicit Burn" events

### PR DESCRIPTION
**Pull request by syuan100**
Discord contact: @alphamethod#0541

## Purpose

There is currently no way to differentiate if a fee was paid using an implicit burn, or with DCs available in an account. This feature aims to expose a method to store and query an implicit burn event by its corresponding transaction hash.

Implementing this feature will make account reconciliation a lot easier, subsequently speeding the Rosetta API implementation along.

## Implementation
I really did not want to touch the base transaction type. Instead, I opted to store Implicit Burn events in the ledger.db in a new column family `ImplicitBurnCF`. This allows the transaction object to be untouched, and implicit burn events can be queried on demand. The key code path to note is the method `debit_fee_from_account` in blockchain_ledger_v1.erl. This is the method called after the HNT to DC conversion rate is calculated and the fee is to be debited from the issuing account. 

At the core, an implicit burn record is stored with only the parameter `fee` denoting the fee in bones. The key is the transaction hash since this is assumed to be an atomic event. 

In order to complete the above, I made the following changes:
- Extended `debit_fee_from_account` to include transaction hash as an optional parameter
- Added blockchain_ledger_implicit_burn_v1.erl to handle the new events
- Changed all instances of `debit_fee_from_account` to include txn hash

## Related PRs
https://github.com/helium/blockchain-node/pull/16
Since the Rosetta implementation will be using blockchain-node, I have made the corresponding changes there to expose a JSONRPC endpoint.

## Questions and concerns
- I'm open to different implementations and would like to understand if there is another angle that should be considered.
- Should this be placed in transactions.db instead of ledger.db? I thought it made more sense in ledger because of where this change is documented in the code path
- Testing is not my strong suit. Are there any tests that I missed or should consider writing?

### CODE REVIEW

**Comment by abhay** [comment]
Thanks for the submission! We'll have a look.

**Comment by evanmcc** [comment]
Thanks so much for the contribution.  While it's quite far along and looks correct so far, I'm not 100% certain we want to bear the cost of adding this information to the ledger at this time.   Let me talk to the rest of the team and we'll see if there is some way to do this, either purely in `blockchain-node` or using some other mechanism that doesn't require us to persist this data forever.

A thing that would help us think it through would be some worked examples of when this data would be used in a larger application.

**Comment by syuan100** [comment]
Hi @evanmcc and thank you for taking a look. I do agree it's some overhead, and perhaps for most applications not necessary. I can look into make this something that gets processed only in blockchain-node, or maybe have it be behind a flag.

The main reason for this feature request is to help with account reconciliation for completing the Rosetta API implementation. Details here: https://www.rosetta-api.org/docs/all_balance_changing.html

Currently, the transaction to_json returns the fee only in DC without any mention of whether or not it was part of a burn. In addition, I haven't found an easy way to tie a transaction to the specific ledger entry that indicates a burn.

Any suggestions are welcome.

**Comment by evanmcc** [comment]
We had a long conversation about this at the end of the day yesterday.  I can provide line-level comments later, but structural changes we'd request:

- This doesn't belong in the ledger.  Please move it to the blocks DB.  Make sure that it's only written on active ledger absorb.  the code would look more like block and snapshot storage, which lives in blockchain.erl
- Make writing this controllable via an application env setting, which should default to false.
- implicit burn records should contain both the fee and the payer address.
- you can probably use some simpler encoding than the verbose ledger datum format if you want, but we need to at least rename it since it isn't going in the ledger (should also move to blockchain/src/ imo).

At that point, one would be able to:
1. run to_json on a txn, grab the hash and the fee
2. check that the fee is not 0
3. if it is, then do `blockchain:get_implicit_burn(Hash, Chain)` to see if it was an implicit burn or not.

Let me know if you have any questions.  Happy to help get this landed.

**Comment by syuan100** [comment]
Thanks for the direction! I'll take another crack at it with that structure in mind and update this PR accordingly.

File: src/ledger/v1/blockchain_ledger_v1.erl [hunk 1]
```diff
@@ -2092,22 +2092,27 @@ debit_account(Address, Amount, Nonce, Ledger) ->
...
+-spec debit_fee_from_account(libp2p_crypto:pubkey_bin(), integer(), ledger(), blockchain_txn:hash(), blockchain:blockchain()) -> ok | {error, any()}.
+debit_fee_from_account(Address, Fee, Ledger, TxnHash, Chain) ->
     case ?MODULE:find_entry(Address, Ledger) of
         {error, _}=Error ->
             Error;
         {ok, Entry} ->
             Balance = blockchain_ledger_entry_v1:balance(Entry),
             case (Balance - Fee) >= 0 of
                 true ->
+                    ImplicitBurn1 = blockchain_implicit_burn:new(
```
Comment (evanmcc) [inline src/ledger/v1/blockchain_ledger_v1.erl:2104]: the whole section here needs to be guarded behind some sort of environment variable.

**Review by evanmcc — COMMENTED** [review]
I think in terms of tests, you should be able to enable this code before any of the add_gateway tests and then check for implicit burns.  Conversely, you should be able check for implicit burns when it is not enabled and not see anything.  Let me know if you need some guidance about which tests to work from.

File: src/blockchain.erl [hunk 2]
```diff
@@ -1727,6 +1732,29 @@ find_last_snapshots(Blockchain, Count0) ->
...
+        Error ->
+            Error
+    end.
+
+-spec add_implicit_burn(blockchain_txn:hash(), blockchain_implicit_burn:implicit_burn(), blockchain()) -> ok | {error, any()}.
+add_implicit_burn(TxnHash, ImplicitBurn, #blockchain{db=DB, implicit_burns=ImplicitBurnsCF}) ->
+    try
+        {ok, Batch} = rocksdb:batch(),
+        {ok, BinImp} = blockchain_ledger_snapshot_v1:serialize(ImplicitBurn),
+        ok = rocksdb:batch_put(Batch, ImplicitBurnsCF, TxnHash, BinImp),
```
[the remark this answers is not in the document]
Comment (syuan100) [reply src/blockchain.erl]: Updated

File: src/blockchain.erl [hunk 3]
```diff
@@ -1756,7 +1784,7 @@ load(Dir, Mode) ->
     case open_db(Dir) of
         {error, _Reason}=Error ->
             Error;
-        {ok, DB, [DefaultCF, BlocksCF, HeightsCF, TempBlocksCF, PlausibleBlocksCF, SnapshotCF]} ->
+        {ok, DB, [DefaultCF, BlocksCF, HeightsCF, TempBlocksCF, PlausibleBlocksCF, ImplicitBurnsCF, SnapshotCF]} ->
```
[the remark this answers is not in the document]
Comment (syuan100) [reply src/blockchain.erl]: Yep fixed! Man I wish I checked here before I started debugging again haha

File: src/ledger/v1/blockchain_ledger_v1.erl (hunk 1 above)
Comment (syuan100) [reply src/ledger/v1/blockchain_ledger_v1.erl]: Would this env variable be something from the command line only, or should I modify the files in the `/config` directory and source the variable from there?

Comment (syuan100) [reply src/ledger/v1/blockchain_ledger_v1.erl]: NVM I'm reading some erlang documentation for `application:get_env` and I think I get the gist of it. I'll take a crack at it

**Comment by syuan100** [comment]
@evanmcc I believe this is now working to spec. I decided to have a check for an env variable called `store_implicit_burns`. I have a forked version of blockchain-node running this branch as core and it works accordingly.

However, I require a little more guidance on writing the tests. I ran `make test` and I got an error while running (see below). Thus, I can't really deduce how to properly run the tests let alone modify an existing one. Any pointers would be appreciated!

```
===> Performing EUnit tests...
.....................................................................................................................................................................................................................................................................................................
Pending:
  blockchain_txn_create_htlc_v1:is_valid_with_extended_validation_test/0
    %% Unknown error: timeout
  undefined
    %% Unknown error: {blame,[1,1,77,12]}


Finished in ? seconds
298 tests, 0 failures, 5 cancelled
===> Error running tests
make: *** [test] Error 1
```

**Comment by vihu** [comment]
> @evanmcc I believe this is now working to spec. I decided to have a check for an env variable called `store_implicit_burns`. I have a forked version of blockchain-node running this branch as core and it works accordingly.
> 
> However, I require a little more guidance on writing the tests. I ran `make test` and I got an error while running (see below). Thus, I can't really deduce how to properly run the tests let alone modify an existing one. Any pointers would be appreciated!
> 
> ```
> ===> Performing EUnit tests...
> .....................................................................................................................................................................................................................................................................................................
> Pending:
>   blockchain_txn_create_htlc_v1:is_valid_with_extended_validation_test/0
>     %% Unknown error: timeout
>   undefined
>     %% Unknown error: {blame,[1,1,77,12]}
> 
> 
> Finished in ? seconds
> 298 tests, 0 failures, 5 cancelled
> ===> Error running tests
> make: *** [test] Error 1
> ```

Ugh, the eunits in erlang aren't _really_ meant to handle long-ish running tests, we need to update these at some point. However, to bypass this error you can do https://erlang.org/doc/apps/eunit/chapter.html#Writing_test_generating_functions

So as an example:

```
%% specify that this test can run for 30s
some_eunit_failing_with_blame() ->
    {timeout,
     30000,
     fun() ->
             %% run the test here
     end
    }.
```

File: src/blockchain.erl [hunk 4]
```diff
@@ -1727,6 +1732,27 @@ find_last_snapshots(Blockchain, Count0) ->
             lists:reverse(List)
     end.
 
+-spec get_implicit_burn(string(), blockchain()) -> {ok, blockchain_implicit_burn:implicit_burn()} | {error, any()}.
```
Comment (vihu) [inline src/blockchain.erl:1735]:
```suggestion
-spec get_implicit_burn(blockchain_txn:hash(), blockchain()) -> {ok, blockchain_implicit_burn:implicit_burn()} | {error, any()}.
```

File: src/blockchain.erl [hunk 5]
```diff
@@ -1727,6 +1732,27 @@ find_last_snapshots(Blockchain, Count0) ->
...
+    case rocksdb:get(DB, ImplicitBurnsCF, TxnHash, []) of
+        {ok, Bin} ->
+            {ok, blockchain_implicit_burn:deserialize(Bin)};
+        not_found ->
+            {error, not_found};
+        Error ->
+            Error
+    end.
+
+-spec add_implicit_burn(string(), blockchain_implicit_burn:implicit_burn(), blockchain()) -> ok | {error, any()}.
```
Comment (vihu) [inline src/blockchain.erl:1746]:
```suggestion
-spec add_implicit_burn(blockchain_txn:hash(), blockchain_implicit_burn:implicit_burn(), blockchain()) -> ok | {error, any()}.
```

File: src/ledger/v1/blockchain_ledger_v1.erl [hunk 6]
```diff
@@ -2092,22 +2092,32 @@ debit_account(Address, Amount, Nonce, Ledger) ->
...
+-spec debit_fee_from_account(libp2p_crypto:pubkey_bin(), integer(), ledger(), blockchain_txn:hash(), blockchain:blockchain()) -> ok | {error, any()}.
+debit_fee_from_account(Address, Fee, Ledger, TxnHash, Chain) ->
     case ?MODULE:find_entry(Address, Ledger) of
         {error, _}=Error ->
             Error;
         {ok, Entry} ->
             Balance = blockchain_ledger_entry_v1:balance(Entry),
             case (Balance - Fee) >= 0 of
                 true ->
+                    case application:get_env(blockchain, store_implicit_burns, false) of
```
Comment (evanmcc) [inline src/ledger/v1/blockchain_ledger_v1.erl:2104]: probably also need to check that `TxnHash` and `Chain` are not `undefined`?

**Comment by vihu** [comment]
@syuan100 since I don't have perms to your fork, [here](https://gist.github.com/vihu/d9901516f19e22aedefe4d02cea8e182) is a simple common test for implicit burn. 

You should be able to run it like so: `./rebar3 ct --suite=blockchain_implicit_burn_SUITE --verbose`

Basically it sets up a payment-v2 transaction with fee enabled and the app env for storing implicit burns set to true. Note that I only have one test there but you should be able to extend it and get some more coverage.

File: src/ledger/v1/blockchain_ledger_v1.erl (hunk 6 above)
Comment (syuan100) [reply src/ledger/v1/blockchain_ledger_v1.erl:2104]: I created a guard condition, but unsure if it is best practice.

**Comment by syuan100** [comment]
@vihu @evanmcc added a few basic tests, let me know if there are other cases to consider

**Comment by syuan100** [comment]
Hello wanted to bump this to see if there were any more comments on this.

I did want to point out that after merging the latest master branch in, there have been a few failing tests. Seems that it has something to do with `dc_entry_not_found`? Not sure how critical this is. See output below:

```
=== Ended at 2021-05-05 21:34:13
=== Location: [{blockchain_simple_SUITE,'-fees_since_test/1-fun-4-',949},
              {lists,foreach,1338},
              {blockchain_simple_SUITE,fees_since_test,943},
              {test_server,ts_tc,1754},
              {test_server,run_test_case_eval1,1263},
              {test_server,run_test_case_eval,1195}]
=== === Reason: no match of right hand side value 
                 {error,
                     {invalid_txns,
                         [{{blockchain_txn_payment_v1_pb,
                               <<0,254,115,61,76,204,191,253,184,43,255,34,186,
                                 226,38,193,118,65,253,94,147,85,75,142,200,55,
                                 24,224,168,249,164,125,252>>,
                               <<0,114,222,130,159,46,150,212,111,205,96,67,70,
                                 88,49,252,213,58,88,123,227,198,225,89,139,
                                 232,119,247,244,103,15,255,88>>,
                               1,10,1,
                               <<48,69,2,32,7,2,39,5,155,167,154,217,242,56,
                                 87,193,6,175,15,235,78,142,168,158,95,83,131,
                                 43,7,180,96,75,10,233,35,70,2,33,0,214,170,
                                 228,142,227,114,95,189,41,148,142,44,182,39,
                                 236,76,243,168,242,74,243,170,169,65,55,203,
                                 193,138,73,195,149,133>>},
                           dc_entry_not_found}]}}
  in function  blockchain_simple_SUITE:'-fees_since_test/1-fun-4-'/6 (/Users/steven.yuan/dev/blockchain-core/test/blockchain_simple_SUITE.erl, line 949)
  in call from lists:foreach/2 (lists.erl, line 1338)
  in call from blockchain_simple_SUITE:fees_since_test/1 (/Users/steven.yuan/dev/blockchain-core/test/blockchain_simple_SUITE.erl, line 943)
  in call from test_server:ts_tc/3 (test_server.erl, line 1754)
  in call from test_server:run_test_case_eval1/6 (test_server.erl, line 1263)
  in call from test_server:run_test_case_eval/9 (test_server.erl, line 1195)
```

```
=== Ended at 2021-05-05 21:34:21
=== Location: [{blockchain_simple_SUITE,routing_test,1023},
              {test_server,ts_tc,1754},
              {test_server,run_test_case_eval1,1263},
              {test_server,run_test_case_eval,1195}]
=== === Reason: no match of right hand side value 
                 {error,
                     {invalid_txns,
                         [{{blockchain_txn_oui_v1_pb,
                               <<0,168,214,244,26,165,106,219,203,101,58,81,
                                 189,150,224,99,157,211,145,80,13,161,45,92,
                                 169,35,55,137,70,170,113,220,18>>,
                               [<<0,245,210,29,14,1,168,229,74,190,171,171,168,
                                  121,251,27,89,120,98,41,202,223,4,235,192,
                                  228,32,216,224,44,22,34,93>>,
                                <<0,57,72,184,57,217,181,56,32,8,150,241,201,
                                  207,33,246,225,34,115,96,113,140,135,96,219,
                                  196,239,153,229,215,199,97,23>>],
                               <<193,92,2,137,236,45,10,145,10,0,0,0,0,0,0,0,0,
                                 0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
                                 0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
                                 0,0,0,0,0,0,0,0,0,0,0,0,0>>,
                               8,<<>>,1,0,
                               <<48,70,2,33,0,231,121,234,253,108,162,79,212,
                                 78,241,23,238,65,100,96,89,32,232,199,139,49,
                                 206,226,6,207,226,75,164,254,57,139,215,2,33,
                                 0,235,219,35,112,31,132,114,179,83,132,211,
                                 200,240,72,68,214,235,13,50,52,141,100,71,74,
                                 232,76,175,147,2,218,56,218>>,
                               <<>>,1},
                           dc_entry_not_found}]}}
  in function  blockchain_simple_SUITE:routing_test/1 (/Users/steven.yuan/dev/blockchain-core/test/blockchain_simple_SUITE.erl, line 1023)
  in call from test_server:ts_tc/3 (test_server.erl, line 1754)
  in call from test_server:run_test_case_eval1/6 (test_server.erl, line 1263)
  in call from test_server:run_test_case_eval/9 (test_server.erl, line 1195)
```

```
=== Ended at 2021-05-05 21:35:32
=== Location: [{blockchain_simple_SUITE,max_subnet_test,1216},
              {test_server,ts_tc,1754},
              {test_server,run_test_case_eval1,1263},
              {test_server,run_test_case_eval,1195}]
=== === Reason: no match of right hand side value 
                 {error,
                     {invalid_txns,
                         [{{blockchain_txn_oui_v1_pb,
                               <<0,107,98,126,147,96,109,165,157,184,227,188,
                                 48,31,95,118,134,179,20,178,227,33,196,61,1,
                                 137,13,169,88,193,120,242,240>>,
                               [<<0,251,116,99,211,136,37,35,6,64,47,35,234,
                                  81,177,85,97,48,167,157,154,122,135,76,252,
                                  189,189,201,253,234,108,165,229>>],
                               <<193,92,2,137,236,45,10,145,10,0,0,0,0,0,0,0,0,
                                 0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
                                 0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
                                 0,0,0,0,0,0,0,0,0,0,0,0,0>>,
                               8,<<>>,1,0,
                               <<48,70,2,33,0,190,8,247,167,12,188,232,101,93,
                                 157,190,109,37,146,71,92,87,37,40,115,73,247,
                                 251,122,147,68,122,2,50,78,36,163,2,33,0,137,
                                 63,0,175,234,139,182,126,144,113,123,220,219,
                                 170,34,63,148,106,219,160,108,58,137,44,152,
                                 219,101,54,92,109,104,35>>,
                               <<>>,1},
                           dc_entry_not_found}]}}
  in function  blockchain_simple_SUITE:max_subnet_test/1 (/Users/steven.yuan/dev/blockchain-core/test/blockchain_simple_SUITE.erl, line 1216)
  in call from test_server:ts_tc/3 (test_server.erl, line 1754)
  in call from test_server:run_test_case_eval1/6 (test_server.erl, line 1263)
  in call from test_server:run_test_case_eval/9 (test_server.erl, line 1195)
```

**Comment by evanmcc** [comment]
Thanks for the bump.  We'll try to do a final review today or tomorrow and merge this.

**Comment by vihu** [comment]
@syuan100 Can you please address the conflicts? We landed a long pending feature in blockchain-core which likely resulted in the conflicts. Once this is ready, we plan on getting this merged :)

**Comment by vihu** [comment]
> @vihu I've fixed the merge conflicts, but `make test` seems to fail on any test where `find_dc_entry` is called still (see above).
> 
> Let me know if this is ok/known issue and if there's something I should do to mitigate. I wonder if it's something to do with my test environment. I would be curious to see if you get the same errors if you pull my branch and run the tests on your end.

Thanks for resolving the conflicts, I’ll take a look at the failing tests shortly.

**Comment by vihu** [comment]
@syuan100 [here's](https://gist.github.com/vihu/148cdff2b04ed11cd5037726008df610) another patch which fixes dialyzer errors. It should apply cleanly with `git apply patch.diff`

**Comment by syuan100** [comment]
@vihu added the patch! Tests have passed. :)

**Comment by syuan100** [comment]
Gentle bump 🙂 I know there's been a lot going on, so no rush!

**Comment by vihu** [comment]
> Gentle bump slightly_smiling_face I know there's been a lot going on, so no rush!

Thank you for the reminder. The team has been tackling some critical blockchain issues and this took a backseat. Anyway, [here's](https://gist.github.com/vihu/ff4b8050efc0a1d10068542eb66f32d2) my final patch to this PR, please apply and confirm that the tests indeed pass for you as well and we'll get this landed!

**Comment by syuan100** [comment]
@vihu Patch applied and tests have passed :) Thank you!

File: test/blockchain_implicit_burn_SUITE.erl [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+    %% Check that there is NO implicit burn
+    {error, not_found} = blockchain:get_implicit_burn(TxHash, Chain),
+    ct:pal("Confirmed that implicit burn has not been stored for ~p", [TxHash]),
+    ok.
+
+%%--------------------------------------------------------------------
+%% HELPERS
+%%--------------------------------------------------------------------
+extra_vars(_) ->
+    #{?txn_fees => true, ?max_payments => ?MAX_PAYMENTS, ?allow_zero_amount => false}.
```
Comment (vihu) [inline test/blockchain_implicit_burn_SUITE.erl:190]:
```suggestion
    #{?txn_fees => true, ?max_payments => ?MAX_PAYMENTS, ?allow_zero_amount => false, ?txn_fee_multiplier => 5000}.
```

**Comment by syuan100** [comment]
Applied the suggestion and tests have passed

**Review by vihu — APPROVED** [review]

**Comment by syuan100** [comment]
Gentle bump! Merged latest master, tests have passed 👍

**Comment by vihu** [comment]
> Gentle bump! Merged latest master, tests have passed +1

Thank you, I have approved this PR and we'll land this tomorrow morning given that a blockchain resync (past ~20K blocks or so) with this works ok (I will start an overnight run today).

**Comment by syuan100** [comment]
Thanks all!! :D

**Pull request merged**